cpp/src/interop/arrow_data_structures.cpp

```diff
@@ -0,0 +1,347 @@
+unique_column_view_t arrow_column::view(rmm::cuda_stream_view stream,
+                                        rmm::device_async_resource_ref mr)
+{
+  return from_arrow_device_column(&container->schema, &container->owner, stream, mr);
+}
+
+arrow_table::arrow_table(cudf::table&& input,
+                         cudf::host_span<column_metadata const> metadata,
+                         rmm::cuda_stream_view stream,
+                         rmm::device_async_resource_ref mr)
+  : container{[&]() {
+      auto schema = cudf::to_arrow_schema(input.view(), metadata);
+      return std::make_shared<arrow_array_container>(
+        std::move(*schema), std::move(input), stream, mr);
+    }()}
+{
+}
+
+unique_table_view_t arrow_table::view(rmm::cuda_stream_view stream,
+                                      rmm::device_async_resource_ref mr)
+{
+  return from_arrow_device(&container->schema, &container->owner, stream, mr);
+}
+
+void arrow_table::to_arrow_schema(ArrowSchema* output,
+                                  rmm::cuda_stream_view stream,
+                                  rmm::device_async_resource_ref mr)
+{
+  ArrowSchemaDeepCopy(&container->schema, output);
+}
+
+void arrow_table::to_arrow(ArrowDeviceArray* output,
+                           ArrowDeviceType device_type,
+                           rmm::cuda_stream_view stream,
+                           rmm::device_async_resource_ref mr)
+{
+  arrow_obj_to_arrow(*this, container, output, device_type, stream, mr);
+}
+
+arrow_table::arrow_table(ArrowSchema&& schema,
+                         ArrowDeviceArray&& input,
+                         rmm::cuda_stream_view stream,
+                         rmm::device_async_resource_ref mr)
+{
+  switch (input.device_type) {
+    case ARROW_DEVICE_CPU: {
+      // I'm not sure if there is a more efficient approach than doing this
+      // back-and-forth conversion without writing a lot of bespoke logic. I
+      // suspect that the overhead of the memory copies will dwarf any extra
+      // work here, but it's worth benchmarking to be sure.
+      auto tbl       = from_arrow_host(&schema, &input, stream, mr);
+      auto tmp_table = arrow_table(std::move(*tbl), get_table_metadata(tbl->view()), stream, mr);
+      container      = tmp_table.container;
+      if (input.array.release != nullptr) { ArrowArrayRelease(&input.array); }
+      break;
+    }
+    default:
+      container =
+        std::make_shared<arrow_array_container>(std::move(schema), std::move(input), stream, mr);
+  }
+}
+
+arrow_table::arrow_table(ArrowSchema&& schema,
+                         ArrowArray&& input,
+                         rmm::cuda_stream_view stream,
+                         rmm::device_async_resource_ref mr)
+{
+  ArrowDeviceArray arr{.array = {}, .device_id = -1, .device_type = ARROW_DEVICE_CPU};
+  ArrowArrayMove(&input, &arr.array);
+  auto tmp  = arrow_table(std::move(schema), std::move(arr), stream, mr);
+  container = tmp.container;
```

Comment:
```suggestion
  container  = arrow_table(std::move(schema), std::move(arr), stream, mr).container;
```
I assume the `arrow_table` instance is destroyed here but the `container` is preserved since it is a shared-ptr.
Is the `tmp` required here?

Reply:
No, it is not. Your change should be fine.